forte/processors/base/data_augment_processor.py

```diff
@@ -46,44 +55,559 @@ class ReplacementDataAugmentProcessor(BaseDataAugmentProcessor):
+        Args:
+            - replacement_op: The class for data augmentation algorithm.
+                For insertion, it should be able to generate a string
+                with an empty input.
+            - data_pack: The datapack for insertion.
+            - pos: The position(index) of insertion.
+        Returns:
+            A bool value. True if the insertion happened, False otherwise.
         """
-        replaced_text: str = replacement_op.replace(input)
-        self.replaced_spans.append((input, replaced_text))
+        pid: int = data_pack.meta.pack_id
+        if self._is_span_overlap(pid, pos, pos):
+            return False
+
+        if pos not in self.inserted_annos[pid]:
+            self.replaced_annos[pid].append((Span(pos, pos), inserted_text))
+            self.inserted_annos[pid][pos] = len(inserted_text)
+            return True
+        return False
+
+    def delete(
+            self,
+            input: Annotation
+    ) -> bool:
+        r"""
+        This is a wrapper function to delete an annotation.
+
+        Args:
+            -input: The annotation to remove.
+        Returns:
+            A bool value. True if the deletion happened, False otherwise.
+        """
+        pid: int = input.pack.meta.pack_id
+        self.replaced_annos[pid].append((input.span, ""))
+        self.deleted_annos[pid].add(input.tid)
+        return True
 
     def auto_align_annotations(
         self,
         data_pack: DataPack,
-        replaced_annotations: List[Tuple[Annotation, str]]
+        replaced_annotations: List[Tuple[Span, str]],
+
     ) -> DataPack:
         r"""
         Function to replace some annotations with new strings.
-        It will update the text and auto-align the annotation spans.
+        It will copy and update the text of datapack and
+        auto-align the annotation spans.
+
+        The links are also copied if its parent & child are
+        both present in the new pack.
 
         Args:
-            data_pack: Datapack holding the annotations to be replaced.
-            replaced_annotations: A list of tuples(annotation, new string).
-            The text for annotation will be updated with the new string.
+            - data_pack: The Datapack holding the replaced annotations.
+            - replaced_annotations: A list of tuples(span, new string).
+            The text and span of the annotations will be updated
+            with the new string.
 
         Returns:
             A new data_pack holds the text after replacement. The annotations
             in the original data pack will be copied and auto-aligned as
-            instructed by the "other_entry_policy".
+            instructed by the "other_entry_policy" in the configuration.
+        """
+        if len(replaced_annotations) == 0:
+            return deepcopy(data_pack)
+
+        # Sort the annotations by span beginning index.
+        replaced_annotations = sorted(
```

Comment:
Why don't why maintained a sorted list during insertion? You can use `SortedList` https://github.com/asyml/forte/blob/master/setup.py#L25